### DISCUSSION TITLE
improve data fetching strategy

### DISCUSSION THREAD

**Discussion by TkDodo** [plausible/analytics#1278] [Anything Else]
Hey, thanks for your cool product 🚀 

I've been looking at the dashboard today, and as a react-query maintainer, I can't unsee the bunch of loading spinners that are popping up every time I filter for some data. I also cannot see when the dashboard was last refreshed / how accurate the data is (which makes me hard-reload the page a lot). I only found out by looking at the code that there seems to be a 30 second refetch interval. Looking at the network tab, I can also see that requests are going out to the same resource multiple times (like e.g. 6 requests going out at the same time to retrieve the current-visitors!)

I think react-query would be a good fit for such a dashboard. Features like `refetchInterval` and window-focus-refetching would make sure that data is always up-to-date, and the stale-while-revalidate caching strategy, paired with background loading indicators, would make sure that users would always see data, which makes the whole dashboard feel snappier. Requests going out at the same time would also be de-duplicated, leading to less server load.

Let me know if you're interested and if / how I can contribute to this 👋

**Comment by ukutaht** [comment]
Hi @TkDodo!

Yes this is definitely something we could improve upon. Everything in your comment is dead on, these are real problems on our dashboard. I guess it's just not hasn't been a big enough problem to make me tackle it properly yet.

I hadn't heard of react-query but it looks really useful for our dashboard. Seems a lot easier than hand-rolling this stuff :)

I will keep this library in mind for future reference. Happy to merge a PR that adds react-query but it will be a pretty big job so up to you if you're interested in contributing @TkDodo.

**Reply by TkDodo** [reply]
Yeah, it would be pretty big. A gradual adoption would also be possible - start off with one request being migrated, followed by another. I can definitely see if I can get you started on this with a PR. Can't promise anything right away, times are pretty busy 😅

**Reply by ukutaht** [reply]
Yeah, a gradual approach might be a better way to go about it. 

> Can't promise anything right away, times are pretty busy sweat_smile

No worries, I'm not expecting anything :)
